torch/csrc/jit/script/init.cpp

```diff
@@ -316,6 +316,17 @@ std::shared_ptr<SugaredValue> toSugaredValue(
   } else if (py::isinstance<py::module>(obj)) {
     return std::make_shared<PythonModuleValue>(obj);
   }
+
+  bool is_weak = py::hasattr(obj, "_jit_is_weak_script") &&
+      py::cast<bool>(py::getattr(obj, "_jit_is_weak_script"));
+
+  if (is_weak) {
+    auto compiled_fn =
+        py::module::import("torch.jit").attr("script")(obj, true, 0, true);
```

Comment:
Hmm shouldn't we compile weak script functions when they are first decorated? Will the Python value lookup work if they will close over some values from their outer scope? Can you try adding that to the test?

Reply:
If we annotate stuff in `torch` with this then that would end up compiling a bunch of graphs that may or may not be used, whereas this only compiles ones that are used in a script as necessary.

Variable closing works the same as in strong scripts (i.e. will throw an error for non-constant values)

Reply:
Yeah, I told David to go with the lazy design so that we feel ok using it across our standard library without incurring startup overhead. It does have weird variable capture problems, but weak_script will at least start out as an internal annotation, and we can avoid variable capture problems. When we make it a public API, we can figure out how to make sure it works fast enough to use it everywhere.